pulser-core/pulser/waveforms.py

```diff
@@ -181,11 +189,22 @@ def modulated_samples(
         Returns:
             The array of samples after modulation.
         """
+        detach = True  # We detach unless...
+        if self.samples.is_tensor and self.samples.as_tensor().requires_grad:
+            # ... the samples require grad. In this case, we clear the cache
+            # so that the modulation is recalculated with the current samples
+            self._modulated_samples.cache_clear()
```

Comment:
The modulated buffers don't have to be recomputed because in differentiation, the duration is not expected to a variable in the gradient (only the values are used), do I get it right ? 😄